src/sendCommercialMetrics.ts

```diff
@@ -1,72 +1,186 @@
 import { log } from '@guardian/libs';
 import { EventTimer } from './EventTimer';
 
-type Metrics = {
+type Metric = {
 	name: string;
 	value: number;
 };
 
-type Properties = {
+type Property = {
 	name: string;
-	value: string;
+	value: string | number;
 };
 
-type CommercialMetrics = {
-	browser_id?: string;
-	page_view_id: string;
-	platform: string;
-	metrics: readonly Metrics[];
-	properties: readonly Properties[];
+type TimedEvent = {
+	name: string;
+	ts: number;
+};
+
+type EventProperties = {
+	type?: ConnectionType;
+	downlink?: number;
+	effectiveType?: string;
+};
+
+type CommercialMetricsPayload = {
+	page_view_id: string | null;
+	browser_id?: string | null;
+	platform: 'NEXT_GEN' | null;
+	metrics: readonly Metric[] | null;
+	properties: readonly Property[] | null;
+};
+
+enum Endpoints {
+	CODE = '//performance-events.code.dev-guardianapis.com/commercial-metrics',
+	PROD = '//performance-events.guardianapis.com/commercial-metrics',
+}
+
+const commercialMetricsPayload: CommercialMetricsPayload = {
+	page_view_id: null,
+	browser_id: null,
+	platform: 'NEXT_GEN',
+	metrics: null,
+	properties: null,
+};
+
+let initialised = false;
+
+const getEndpoint = (isDev: boolean) => {
+	return isDev ? Endpoints.CODE : Endpoints.PROD;
+};
+
+const getDevProperties = (isDev: boolean): Property[] => {
+	return isDev ? [{ name: 'isDev', value: window.location.hostname }] : [];
+};
+
+const getAdBlockerProperties = (adBlockerInUse?: boolean): Property[] => {
+	return adBlockerInUse !== undefined
+		? [
+				{
+					name: 'adBlockerInUse',
+					value: adBlockerInUse.toString(),
+				},
+		  ]
+		: [];
+};
+
+const filterUndefinedEventTimerProperties = (
+	eventTimerProperties: EventProperties,
+): Array<[string, string | number]> => {
+	// Transforms object {key: value} pairs into an array of [key, value] arrays
+	return Object.entries(eventTimerProperties).filter(
+		([, value]) => typeof value !== 'undefined',
+	);
+};
+
+const mapEventTimerPropertiesToString = (
+	properties: Array<[string, string | number]>,
+): Property[] => {
+	return properties.map(([name, value]) => ({
+		name: String(name),
+		value: String(value),
+	}));
+};
```

Comment:
We already have an object cleaner in ad-targeting:  https://github.com/guardian/commercial-core/blob/main/src/targeting/pick-targeting-values.ts.

If this is something we end up doing a lot of, maybe we’d want to have a general utility to remove undefined keys from an object.

Reply:
Yes, that's a good point. Do you think it's at that point now? Or just something to keep in mind?

Reply:
I think just keep in mind at this point 😁